presto-native-sidecar-plugin/src/main/java/com/facebook/presto/sidecar/nativechecker/NativePlanChecker.java

```diff
@@ -0,0 +1,250 @@
+            return new URL(nodeManager.getSidecarNode().getHttpUri().toString() + PLAN_CONVERSION_ENDPOINT);
+        }
+        catch (IOException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "I/O Error opening native sidecar location", e);
+        }
+    }
+
+    @Override
+    public void validate(PlanNode planNode, WarningCollector warningCollector)
+    {
+        // NO-OP, only validating fragments
+    }
+
+    @Override
+    public void validateFragment(SimplePlanFragment planFragment, WarningCollector warningCollector)
+    {
+        if (!planFragment.getPartitioning().isCoordinatorOnly() && !isInternalSystemConnector(planFragment.getRoot())) {
+            runValidation(planFragment);
+        }
+        else {
+            LOG.debug("Skipping native plan validation for plan fragment id: %s", planFragment.getId());
+        }
+    }
+
+    private boolean isInternalSystemConnector(PlanNode planNode)
+    {
+        return planNode.accept(new CheckInternalVisitor(), null);
+    }
+
+    private void runValidation(SimplePlanFragment planFragment)
+    {
+        LOG.debug("Starting native plan validation for plan fragment id: %s", planFragment.getId());
+        String requestBodyJson = planFragmentJsonCodec.toJson(planFragment);
+        final Request request = buildRequest(requestBodyJson);
+
+        try (Response response = httpClient.newCall(request).execute()) {
+            if (!response.isSuccessful()) {
+                PlanConversionFailure failure = processResponseFailure(response);
+                throw new PrestoException(failure, "Error from native plan checker: " + failure.getMessage());
+            }
+        }
+        catch (final IOException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "I/O error getting native plan checker response", e);
+        }
+        finally {
+            LOG.debug("Native plan validation complete for plan fragment id: %s", planFragment.getId());
+        }
+    }
+
+    private Request buildRequest(String requestBodyJson)
+    {
+        Request.Builder builder = new Request.Builder()
+                .url(getPlanValidateUrl())
+                .addHeader("CONTENT_TYPE", "APPLICATION_JSON")
+                .post(RequestBody.create(JSON_CONTENT_TYPE, requestBodyJson));
+
+        return builder.build();
+    }
+
+    private PlanConversionFailure processResponseFailure(Response response) throws IOException
+    {
+        if (response.body() != null) {
+            PlanConversionResponse planConversionResponse = PLAN_CONVERSION_RESPONSE_JSON_CODEC.fromJson(response.body().bytes());
+            if (!planConversionResponse.getFailures().isEmpty()) {
+                return planConversionResponse.getFailures().get(0);
+            }
+            else {
+                throw new PrestoException(GENERIC_INTERNAL_ERROR, "Error response without failure from native plan checker with code: " + response.code());
```

Comment:
Let's make this a dedicated internal error.   We can make a new instance of ErrorCodeSupplier for this plugin and add this there.